presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.prestosql.operator.exchange.PageChannelSelector;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.spi.predicate.ValueSet;
+import io.prestosql.spi.type.TypeUtils;
+import io.prestosql.sql.planner.DynamicFilter;
+import io.prestosql.sql.planner.plan.PlanNodeId;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Verify.verify;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverByteSize;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverRowCount;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This operator acts as a simple "pass-through" pipe, while saving its input pages.
+ * The collected pages' value are used for creating a run-time filtering constraint (for probe-side table scan in an inner join).
+ * We support only small build-side pages (which should be the case when using "broadcast" join).
+ */
+public class DynamicFilterCollectOperator
+        implements Operator
+{
+    public static class DynamicFilterCollectOperatorFactory
+            implements OperatorFactory, Consumer<Optional<TupleDomain<ColumnHandle>>>
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final DynamicFilter dynamicFilter;
+
+        private TupleDomain<ColumnHandle>[] partitions;
+        private int partitionCount;
+        private final Object partitionLock;
+
+        public DynamicFilterCollectOperatorFactory(int operatorId, PlanNodeId planNodeId, DynamicFilter dynamicFilter, int partitionCount)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.dynamicFilter = requireNonNull(dynamicFilter);
+
+            this.partitions = new TupleDomain[partitionCount];
+            this.partitionLock = new Object();
+        }
+
+        @Override
+        public void accept(Optional<TupleDomain<ColumnHandle>> dynamicPredicate)
```

Comment:
I don't think it's needed. Could you add a method like:
```
Consumer<TupleDomain<ColumnHandle>> DynamicFilter#getPartitionTupleDomainConsumer()
```
On last consumed partition it would do the union and set final future with `TupleDomain` for all partitions.

Reply:
Sounds good, fixed at c343902c15.